### ISSUE TITLE
discordChannel.sendMessage is not a function

### ISSUE THREAD

**Issue by dustinlacewell** [reactiflux/discord-irc#132]
When mapping multiple channels:

```
[
  {
    "nickname": "o",
    "server": "chat.freenode.net",
    "discordToken": "abc",
    "autoSendCommands": [
      ["PRIVMSG", "NickServ", "IDENTIFY abc"]
    ],
    "channelMapping": {
        "#lojban": "#lojban",
        "#tadni": "#ckule"
    }
  }
]
```

I get the following failure as soon as anyone says anything in the IRC channel `#ckule`:

```
TypeError: discordChannel.sendMessage is not a function
    at Bot.sendToDiscord (/usr/local/lib/node_modules/discord-irc/dist/bot.js:204:22)
    at emitMany (events.js:127:13)
    at Client.emit (events.js:201:7)
    at Client.<anonymous> (/usr/local/lib/node_modules/discord-irc/node_modules/irc/lib/irc.js:553:22)
    at emitOne (events.js:96:13)
    at Client.emit (events.js:188:7)
    at iterator (/usr/local/lib/node_modules/discord-irc/node_modules/irc/lib/irc.js:842:26)
    at Array.forEach (native)
    at Socket.handleData (/usr/local/lib/node_modules/discord-irc/node_modules/irc/lib/irc.js:837:15)
    at emitOne (events.js:96:13)
```

It does not seem to matter what the naming of either the discord or IRC channels are.

**Comment by ekmartin** [comment]
What version of the bot are you using? Could you try uninstalling and reinstalling it and see if that helps? There's been a lot of changes in the last versions in terms of Discord support.

I can't reproduce this with a multichannel config using the last version.

**Comment by dustinlacewell** [comment]
The version installed from NPM

**Comment by dustinlacewell** [comment]
When I clone to `/discord-irc` with the config located at `/config.json` and run `/discord-irc# npm start -- --config /config.json` after `npm install` I get:

``` 
npm info it worked if it ends with ok
npm info using npm@3.8.9
npm info using node@v6.2.0
npm info lifecycle discord-irc@2.0.2~prestart: discord-irc@2.0.2
npm info lifecycle discord-irc@2.0.2~start: discord-irc@2.0.2
> discord-irc@2.0.2 start /discord-irc
> node dist/index.js "--config" "/config.json"

module.js:442
    throw err;
    ^

Error: Cannot find module '/discord-irc/dist/index.js'
    at Function.Module._resolveFilename (module.js:440:15)
    at Function.Module._load (module.js:388:25)
    at Function.Module.runMain (module.js:575:10)
    at startup (node.js:160:18)
    at node.js:449:3

npm info lifecycle discord-irc@2.0.2~start: Failed to exec start script
npm ERR! Linux 4.4.0-36-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "start" "--" "--config" "/config.json"
npm ERR! node v6.2.0
npm ERR! npm  v3.8.9
npm ERR! code ELIFECYCLE
npm ERR! discord-irc@2.0.2 start: `node dist/index.js "--config" "/config.json"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the discord-irc@2.0.2 start script 'node dist/index.js "--config" "/config.json"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the discord-irc package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node dist/index.js "--config" "/config.json"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs discord-irc
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls discord-irc
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /discord-irc/npm-debug.log

**Comment by dustinlacewell** [comment]
I was able to get the process started with the following:

```
node ./node_modules/babel-cli/bin/babel.js lib --out-dir dist                                                    
lib/bot.js -> dist/bot.js
lib/cli.js -> dist/cli.js
lib/errors.js -> dist/errors.js
lib/helpers.js -> dist/helpers.js
lib/index.js -> dist/index.js
lib/validators.js -> dist/validators.js
```

Which actually created the `dist` folder, allowing to run, `npm start -- --config /config.json` but then I get the same:

```
npm info it worked if it ends with ok                                                                                                    [8/1839]
npm info using npm@3.8.9
npm info using node@v6.2.0
npm info lifecycle discord-irc@2.0.2~prestart: discord-irc@2.0.2
npm info lifecycle discord-irc@2.0.2~start: discord-irc@2.0.2

> discord-irc@2.0.2 start /discord-irc
> node dist/index.js "--config" "/config.json"

/discord-irc/node_modules/irc/lib/irc.js:845
                        throw err;
                        ^

TypeError: discordChannel.sendMessage is not a function
    at Bot.sendToDiscord (/discord-irc/dist/bot.js:204:22)
    at emitMany (events.js:127:13)
    at Client.emit (events.js:201:7)
    at Client.<anonymous> (/discord-irc/node_modules/irc/lib/irc.js:553:22)
    at emitOne (events.js:96:13)
    at Client.emit (events.js:188:7)
    at iterator (/discord-irc/node_modules/irc/lib/irc.js:842:26)
    at Array.forEach (native)
    at Socket.handleData (/discord-irc/node_modules/irc/lib/irc.js:837:15)
    at emitOne (events.js:96:13)

npm info lifecycle discord-irc@2.0.2~start: Failed to exec start script
npm ERR! Linux 4.4.0-36-generic
npm ERR! argv "/usr/local/bin/node" "/usr/local/bin/npm" "start" "--" "--config" "/config.json"
npm ERR! node v6.2.0
npm ERR! npm  v3.8.9
npm ERR! code ELIFECYCLE
npm ERR! discord-irc@2.0.2 start: `node dist/index.js "--config" "/config.json"`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the discord-irc@2.0.2 start script 'node dist/index.js "--config" "/config.json"'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the discord-irc package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node dist/index.js "--config" "/config.json"
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs discord-irc
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls discord-irc
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /discord-irc/npm-debug.log
```

Here is the `npm-debug.log`:

```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/local/bin/node',
1 verbose cli   '/usr/local/bin/npm',
1 verbose cli   'start',
1 verbose cli   '--',
1 verbose cli   '--config',
1 verbose cli   '/config.json' ]
2 info using npm@3.8.9
3 info using node@v6.2.0
4 verbose run-script [ 'prestart', 'start', 'poststart' ]
5 info lifecycle discord-irc@2.0.2~prestart: discord-irc@2.0.2
6 silly lifecycle discord-irc@2.0.2~prestart: no script for prestart, continuing
7 info lifecycle discord-irc@2.0.2~start: discord-irc@2.0.2
8 verbose lifecycle discord-irc@2.0.2~start: unsafe-perm in lifecycle true
9 verbose lifecycle discord-irc@2.0.2~start: PATH: /usr/local/lib/node_modules/npm/bin/node-gyp-bin:/discord-irc/node_modules/.bin:/usr/local/bin
:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
10 verbose lifecycle discord-irc@2.0.2~start: CWD: /discord-irc
11 silly lifecycle discord-irc@2.0.2~start: Args: [ '-c', 'node dist/index.js "--config" "/config.json"' ]
12 silly lifecycle discord-irc@2.0.2~start: Returned: code: 1  signal: null
13 info lifecycle discord-irc@2.0.2~start: Failed to exec start script
14 verbose stack Error: discord-irc@2.0.2 start: `node dist/index.js "--config" "/config.json"`
14 verbose stack Exit status 1
14 verbose stack     at EventEmitter.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/lifecycle.js:245:16)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at EventEmitter.emit (events.js:191:7)
14 verbose stack     at ChildProcess.<anonymous> (/usr/local/lib/node_modules/npm/lib/utils/spawn.js:24:14)
14 verbose stack     at emitTwo (events.js:106:13)
14 verbose stack     at ChildProcess.emit (events.js:191:7)
14 verbose stack     at maybeClose (internal/child_process.js:850:16)
14 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:215:5)
15 verbose pkgid discord-irc@2.0.2
16 verbose cwd /discord-irc
17 error Linux 4.4.0-36-generic
18 error argv "/usr/local/bin/node" "/usr/local/bin/npm" "start" "--" "--config" "/config.json"
19 error node v6.2.0
20 error npm  v3.8.9
21 error code ELIFECYCLE
22 error discord-irc@2.0.2 start: `node dist/index.js "--config" "/config.json"`
22 error Exit status 1
23 error Failed at the discord-irc@2.0.2 start script 'node dist/index.js "--config" "/config.json"'.
23 error Make sure you have the latest version of node.js and npm installed.
23 error If you do, this is most likely a problem with the discord-irc package,
23 error not with npm itself.
23 error Tell the author that this fails on your system:
23 error     node dist/index.js "--config" "/config.json"
23 error You can get information on how to open an issue for this project with:
23 error     npm bugs discord-irc
23 error Or if that isn't available, you can get their info via:
23 error     npm owner ls discord-irc
23 error There is likely additional logging output above.
24 verbose exit [ 1, true ]
```

This is using ToT master and node v 6.2.0

**Comment by dustinlacewell** [comment]
Figured the installation issues out. `npm install` was failing in a strange way I didn't notice because of the fact that I'm running this in a Docker container and the user inside is root. Adding `--unsafe-perm` worked around the issue.

However, the bot still crashes when text comes from IRC from one of the channels with the original error.

**Comment by ekmartin** [comment]
@dustinlacewell you need to run `npm run build` after installing. I've added it to the README, quite embarassing that it wasn't there already. If you run the application with `NODE_ENV=development` you'll get some more debug logs, could you try that? `NODE_ENV=development npm start -- --config config.json`.

Also, if you have time, try inserting a `console.log(discordChannel);` here: https://github.com/reactiflux/discord-irc/blob/master/lib/bot.js#L189

**Comment by dustinlacewell** [comment]
So I added
```
        if (typeof(discordChannel.sendMessage) != 'function') {
            logger.info('Couldnt send to ', discordChannelName, 'because sendMessage is ', discordChannel.sendMessage);
            return
        }
```

And so now it doesn't crash, but for 1 channel out of the three I map, I get:

`info: Couldnt send to  #jbosnu because sendMessage is  undefined`

And obviously don't get any messages from that IRC channel.

**Comment by ekmartin** [comment]
Could you try logging the actual discordChannel object, not the function? I.e. `console.log(discordChannel)`. It has to be an object, otherwise `discordChannel.sendMessage` would crash with a can't read sendMessage of undefined error.

**Comment by dustinlacewell** [comment]
The message becomes, 

```
info: Couldnt send to  #jbosnu because discordChannel is  type=voice, id=230541181161832448, name=jbosnu, position=2, , bitrate=64000, userLimit=0, , , , available=true, id=230498134843850762, name=lojban, icon=1acf271d61fb7c38d91cfabd7d4ec95f, splash=null, region=us-west, memberCount=22, large=undefined, features=[], , afkTimeout=300, afkChannelID=null, embedEnabled=undefined, verificationLevel=0, _joinedTimestamp=1475033613613, ownerID=127697568166576128, , 
```

Does this indicate that it is grabbing the voice channel and that's why it can't send to it? Strange because we have a voice channel for every text channel. Why does `#jbosnu` the only channel that it grabs the voice channel vs the text channel?

**Comment by dustinlacewell** [comment]
Woot I fixed it. I'll submit a PR.
